kubevirt: VMWizard - redesign and import the first step (General)

suomiy opened the pull request:
depends on:
- [x] #2453
- [x] #2677

- adds form helpers like FormFieldRow
- uses immutables and memoization for performance
- adds EnhancedK8sMethods
- move wizard logic to actions

@mareklibra @rawagner please review

according to the new design
http://openshift.github.io/openshift-origin-design/web-console/knikubevirt/knikubevirt

![pic](https://user-images.githubusercontent.com/3648838/64727627-896b1980-d4d9-11e9-97c4-15831a14e6b6.png)


![err](https://user-images.githubusercontent.com/3648838/63507713-a31bd100-c4d8-11e9-8a93-02cc73df609f.png)

@matthewcarleton

suomiy commented:
/hold 
for 4.3

suomiy commented:
The wizard is not complete yet, but the features and steps will be added incrementally in the future

matthewcarleton commented:
@rawagner this is looking great! Thank you for doing this :)
Here's a few design tweaks I think we could do to better align things with PatternFly:
There should be a next button on the first step. 
Neither the Create and Review button or the Next button should be disabled as they should alert the user of the needed fields before they proceed.
We can remove the Back button from the first step in the same way we remove the Back button in the last step.
The tooltip icon should be `pf-icon-help` and `#72767B` unless active then it should be `#0066CC`
We've dropped the tooltip from the Operating System 
We've renamed `Provision source` to just `Source` to align with the storage step.
The alert you are showing should be updated to use the inline variation [see here](https://www.patternfly.org/v4/documentation/core/components/alert#inline-alert-types)
Is this alert shown after they have tried to proceed? If so does it scroll the user back to the top so they can see it?
If possible, I think we should use input help text(red) rather than larger inline ones if it's referring to a specific input.
Does the container image field need to be a text area? I've also wondered with the fields where they just paste in a path if we should provide helper text below to give some guidance. 
Sorry this got so long :)

suomiy commented:
> @rawagner this is looking great! Thank you for doing this :)
> Here's a few design tweaks I think we could do to better align things with PatternFly:
> There should be a next button on the first step.

There will be in the future, but I put `Review and Create` just for now because this PR adds just the first step.

> Neither the Create and Review button or the Next button should be disabled as they should alert the user of the needed fields before they proceed.

We are doing real-time validation of input, so the user doesn't have click on the button to see the error messages. This way it is more natural to keep in disabled until all required fields are valid.

> The tooltip icon should be `pf-icon-help` and `#72767B` unless active then it should be `#0066CC`
> We've dropped the tooltip from the Operating System (we only have it on Source, Flavor, and Workload profile)
> We've renamed `Provision source` to just `Source` to align with the storage step.
> The alert you are showing should be updated to use the inline variation [see here](https://www.patternfly.org/v4/documentation/core/components/alert#inline-alert-types)

All done!

> Is this alert shown after they have tried to proceed? If so does it scroll the user back to the top so they can see it?
> If possible, I think we should use input help text(red) rather than larger inline ones if it's referring to a specific input.

User will see these alerts when they open the dialog. These are alerts for the resources which could not load correctly. And these might be used by multiple fields or none.

> Does the container image field need to be a text area? I've also wondered with the fields where they just paste in a path if we should provide helper text below to give some guidance.

switched to text area

> Sorry this got so long :)

np, thanks for the comments :) 

current version:

![tu](https://user-images.githubusercontent.com/3648838/63600376-beafd600-c5c3-11e9-9942-420aa4d3374e.png)

matthewcarleton commented:
@suomiy 
+1 on your changes :)
disabled buttons:
Having fields dynamically validate is awesome but I still think these buttons shouldn't be disabled.
My reasoning is mostly because of accessibility. There's lots of scenarios where a user (keyboard only, screen reader, high zoom) will have problems and not understand why. They might miss the announcement of the error, they might be zoomed in and not see it or their vision could be poor and they don't notice it. 
With a disabled button we don't have the opportunity to help them when these scenarios take place. We just assume they will figure it out on their own (which likely they will eventually, or just give up).
Having the button enabled helps them understand what's wrong with the form via an alert when they've tried to submit without success. This puts them in a place where they don't have to problem solve only act on the guidance they're given.
What if we did something like this: 
<img width="638" alt="Screen Shot 2019-08-23 at 1 02 06 PM" src="https://user-images.githubusercontent.com/287880/63606362-44bd2400-c5a6-11e9-9af6-d0a47be1ea11.png">

>User will see these alerts when they open the dialog. These are alerts for the resources which could not load correctly. And these might be used by multiple fields or none.

Is this a real scenario or just a placeholder? I'm assuming in this scenario they've tried to create this VM and ran into this error and are now back in the wizard trying to correct it?  If that's the case wouldn't they be on the results page? From there they would go to the storage step to correct the issue with the storage class so they could see this error on the page. What have I missed :) ?

suomiy commented:
> ...
> What if we did something like this:

I agree, you have good points. I am not tackling the VMWizard navigation & logic much in this PR. 
I will try to do address these issues in the next PR, if that's okay. 

> Is this a real scenario or just a placeholder? I'm assuming in this scenario they've tried to create this VM and ran into this error and are now back in the wizard trying to correct it? If that's the case wouldn't they be on the results page? From there they would go to the storage step to correct the issue with the storage class so they could see this error on the page. What have I missed :) ?

Real scenario. No, they just opened the dialog without filling anything and are presented with these errors; usually permission issues which result in an inability of using the whole wizard.

suomiy commented:
rebased

suomiy commented:
> Ah ok, I wonder then if we should just present them with this error and guidance on how to fix it. I suppose there isn't much value in providing all of the fields if they can't proceed until the resolve the issue. WDYT?

Not sure, because there are some errors which will disable the dialog only partially (e.g. datavolumes). You can still create some configurations without them.

suomiy commented:
/retest

suomiy commented:
/retest

suomiy commented:
fixed issues mentioned in offline review

rawagner commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx:
```diff
@@ -0,0 +1,311 @@
+    }),
+  ];
+
+  const dataIDReferences = resources.reduce((acc, resource) => {
```
firehose injects `reduxIDs` props so you dont have to compute them yourself

suomiy replied:
That is true, but it is not enough. We need to pair the reference path with the prop. Please see:

```tsx
 acc[resource.prop] = ['k8s', makeReduxID(resource.model, query)];
```

This is done to get our desired prop even if the query and redux id changes (e.g. namespace change).

suomiy commented:
/hold cancel

suomiy commented:
/retest

mareklibra commented:
lgtm
waiting for the branch to merge this to

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx:
```diff
@@ -0,0 +1,331 @@
+        activeNamespace={activeNamespace}
+        reduxID={reduxID}
+        onClose={() => history.goBack()}
+        footer={() => <div>test</div>}
```
removed unnecessary prop

note: the full feature footer is implemented in the followup https://github.com/openshift/console/pull/2565

rawagner commented:
/lgtm
